lib/cwrap.c

```diff
@@ -0,0 +1,245 @@
+int pathCheck(char *path){
+	int i;
+	struct stat statbuf;
+
+		if(stat(path,&statbuf)==0){
+			if (S_ISREG (statbuf.st_mode)){
+				i=0;
+			}
+			 if (S_ISDIR (statbuf.st_mode)) {
+			        i=-1;
+			    }
+
+		}
+		return i;
+}
+/* Opens a directory and store both the directoryfd and
+   the directory path in a structure
+*/
+
+opened_dir_struct * open_directory(char* file_path,opened_dir_struct *dos){
+	int dir_fd,k,j; char * dirname;
+	DIR *dir;
+	k=pathCheck(file_path);
+
+	if(k==0){
+		perror("");
+	}
+	else{
+		dirname=file_path;
+	}
+	dir=opendir(dirname);
+
+	if(dir !=NULL){
+		dir_fd=dirfd(dir);
+		dos->dirfd=dir_fd;
+		dos->dirname=dirname;
+
+	}
+	else{
+		perror("");
+	}
+
+	return dos;
+
+}
+
+
+Map* add_Opened_dirpath_map(opened_dir_struct ods){
+	int current=map->length,i;
+	map->opened_files[map->length]=ods;
+	map->length++;
+	return map;
+}
+Map* preopen(char* file,int mode){
+	int k;
+	char* dirname;
+	opened_dir_struct ods;
+	opened_dir_struct * odsp;
+	k=checkCapacity();
+	if(k<0){
+		map=increaseMapCapacity();
+	}
+
+	odsp=open_directory( file,&ods);
+	map=add_Opened_dirpath_map(*odsp);
+	return map;
+}
+
+/*
+  Finds how many characters in a string is in another
+  string begining from the first character
+*/
+int findMatchingChars(char *A,char *B){
+	//argu *A is path to be matched
+	//argu *B is a preopened dir path in the map pointer
+	int lenA,lenB,matchedNum=0,i;
+	assert(A!=NULL && B!=NULL);
+	lenA=strlen(A);
+	lenB=strlen(B);
+	for(i=0;i<lenB;i++){
+		if(A[i]==B[i]){
+			matchedNum++;
+		}
+		else{
+			break;
+		}
+	}
+	return matchedNum;
+}
+
+/*
+ Returns the dirfd of the opened path with highest matched number
+ and relative path to the dirfd
+*/
+int  getMostMatchedPath(int matches[]){
+	int highestnumber=0,i,highest_num_index=0;
+
+	for(i=0;i<map->length;i++){
+		if(matches[i]>highestnumber){
+			highestnumber=matches[i];
+			highest_num_index=i;
+
+		}
+
+	}
+
+	return matches[highest_num_index];
+
+}
+/*compares matched path and see if the matched path is already opened
+ * if not it opens the matched path else it returns the matched path dirfd
+ * and relative path.
+ */
+matched_path compareMatched(Map* map,int best_matched_num,char *newPath,int mode){
+	char * temp_dir,*t_dir,*filename;
+	//const char* slash ="/";
+	int i,status;
+	matched_path  matchedPath ={0};
+	if(best_matched_num==0){
+		map=preopen(newPath,mode);
+		matchedPath.dirfd=map->opened_files[map->length-1].dirfd;
+		matchedPath.relative_path=newPath+strlen(map->opened_files[map->length-1].dirname);
+	}
+	else{
+
+				t_dir=newPath+best_matched_num;
+				temp_dir=strndup(newPath,strlen(newPath)- strlen(t_dir));
+
+				for(i=0;i<map->length;i++){
+					status=strcmp(temp_dir,map->opened_files[i].dirname);
+					if(status==0){
+						matchedPath.dirfd=map->opened_files[i].dirfd;
+						matchedPath.relative_path=t_dir;
+						break;
+					}
+				}
+				if(status !=0){
+					map=preopen(newPath,mode);
```

Comment:
Again, let's not modify the map in this function: let's just look for existing matches. In fact, let's make the map parameter to this function `const` just to make this explicit.